Sources/HummingbirdWSCore/WebSocket.swift

```diff
@@ -12,50 +12,132 @@
 //
 //===----------------------------------------------------------------------===//
 
+import Atomics
 import NIOCore
 import NIOWebSocket
 
 /// WebSocket object
-public final class HBWebSocket {
+public final class HBWebSocket: Sendable {
     public enum SocketType: Sendable {
         case client
         case server
     }
 
+    private class AutoPingTaskManager {
+        private var waitingOnPong: Bool
+        private var pingData: ByteBuffer
+        private var autoPingTask: Scheduled<Void>?
+
+        init(channel: Channel) {
+            self.waitingOnPong = false
+            self.pingData = channel.allocator.buffer(capacity: 16)
+            self.autoPingTask = nil
+            channel.closeFuture.whenComplete { _ in
+                self.shutdown()
+            }
+        }
+
+        func shutdown() {
+            self.autoPingTask?.cancel()
+        }
+
+        /// Send ping and setup task to check for pong and send new ping
+        func initiateAutoPing(interval: TimeAmount, ws: HBWebSocket) {
+            self.autoPingTask = ws.channel.eventLoop.scheduleTask(in: interval) {
+                if self.waitingOnPong {
+                    // We never received a pong from our last ping, so the connection has timed out
+                    let promise = ws.channel.eventLoop.makePromise(of: Void.self)
+                    ws.close(code: .goingAway, promise: promise)
+                    promise.futureResult.whenComplete { _ in
+                        // Usually, closing a WebSocket is done by sending the close frame and waiting
+                        // for the peer to respond with their close frame. We are in a timeout situation,
+                        // so the other side likely will never send the close frame. We just close the
+                        // channel ourselves.
+                        ws.channel.close(mode: .all, promise: nil)
+                    }
+
+                } else {
+                    ws.sendPing().whenSuccess {
+                        self.waitingOnPong = true
+                        self.initiateAutoPing(interval: interval, ws: ws)
+                    }
+                }
+            }
+        }
+
+        /// Respond to pong from client. Verify contents of pong and clear waitingOnPong flag
+        func receivedPong(frame: WebSocketFrame, ws: HBWebSocket) {
+            let frameData = frame.unmaskedData
+            guard frameData == self.pingData else {
+                ws.close(code: .goingAway, promise: nil)
+                return
+            }
+            self.waitingOnPong = false
+            ws.pongCallback.load(ordering: .relaxed).wrapped?(ws)
+        }
+
+        /// Send ping message
+        /// - Parameter promise: promise that is completed when ping message has been sent
+        func sendPing(ws: HBWebSocket, promise: EventLoopPromise<Void>?) {
+            if self.waitingOnPong {
+                promise?.succeed(())
```

Comment:
This doesn't seem technically correct, but I get why it's there. This is the only thing I'd consider updating or making an issue for

Reply:
HBWebSocket only supports one inflight ping. If it is waiting for a pong a ping must have already been sent.